### ISSUE TITLE
Scenario - Shared Dependencies Across Subrepos

### ISSUE THREAD

**Issue by melody-universe** [ingydotnet/git-subrepo#528]
Hi all, starting to ramp up into this project and really enjoy the work you've done!

I've been trying to conceive of an approach using subrepos in which their dependencies can be shared. For a very basic example, I have adopted a pattern in which I maintain `git-subrepo` in my repos, so that I don't need to add manual installation instructions for all the developers I work with. I've created a gist with instructions [here](https://gist.github.com/tyler-hogsett-microsoft/bc4f8fd7e0e240b2e89de7c9b1cb70f9) (please forgive my tight coupling to Windows).

I know `git-subrepo` supports subrepos within subrepos, but I don't want my end users' monorepos to have 18 different copies of `git-subrepo` distributed throughout their monorepo. Instead, it would be great if subrepos could have some sort of file (perhaps reusing `.gitrepo`) that defines symlinks from that subrepo to some folder maintained in the root of my monorepo.

Does this make any sense? I'm sure there are a million gotcha's that I'm missing that might make this a very bad idea, but I wanted to open a discussion with this community to illuminate some of those for me.

The one that comes to mind is potential version discrepancies, so for example, subrepo A depends on subrepo C at commit 1, and subrepo B depends on subrepo C at commit 2. If commit 2 contains breaking changes relative to commit 1 in subrepo C, subrepo A would need to be updated to work with the new functionality. But I believe npm has a way to handle problems like this.

**Comment by admorgan** [comment]
One of the design parameters of subrepo is that it is just a repo that can be used like any other repo. All the subrepo magic is hidden from git and only come into play when pulling or pushing. This design parameter makes things like using the .gitrepo file contain descriptions of symlinks not viable, as it would require some interaction as git would not natively know to populate them.

As far as having a lot of copies of files in a repo, subrepo itself has two copies of the bash+ library, one for subrepo and one for the test infrastructure test-more-bash. This is done explicitly for the dependency issues you describe. NPM handles these problems by nesting versions and managing their applicable scope. Python requires nested virtual environments to handle the issue.

I don't use VSCode enough to know how it handles nested configurations. In Eclipse they would be loaded as individual projects with their own settings and therefore would do the "right thing" and execute the version of subrepo for that code.

I personally like using hooks to handle project configuration. I personally use pre-commit from http://pre-commit.com, or course that also requires you have a tool pre-installed on your desktop.

In summary, having multiple copies of subrepo in your repo probably won't hurt you, but I understand why it is not optimal. Using .gitrepo to do "work" in git is not viable.
